<a href="https://colab.research.google.com/github/mrzack18/Sistem-Pakar-Certainty-Factor-Fuzzy-Logic-/blob/main/CertaintyFactor.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **1. Ubah nilai CF gejala dan amati perubahan hasil**
Misalnya mengubah "demam": 0.3 dari sebelumnya 0.7, "batuk": 0.6 dari sebelumnya 0.5, dan "sakit_tenggorokan": 0.6 dari sebelumnya 0.4 lalu jalankan kembali fungsi hitung_cf.



In [32]:
# Gejala user (demam diubah menjadi 0.3, batuk menjadi 0.6 dan sakit tenggorokan 0.6)
gejala_user = {
    "demam": 0.3,
    "batuk": 0.6,
    "sakit_tenggorokan": 0.6
}

# Pengetahuan Pakar
pengetahuan = {
    "flu": {
        "demam": 0.8,
        "batuk": 0.7,
        "sakit_tenggorokan": 0.6
    }
}

# Fungsi perhitungan CF
def hitung_cf(gejala_user, pengetahuan_pakar):
    cf_total = 0
    first = True
    for gejala, cf_user in gejala_user.items():
        if gejala in pengetahuan_pakar:
            cf_pakar = pengetahuan_pakar[gejala]
            cf = cf_user * cf_pakar
            if first:
                cf_total = cf
                first = False
            else:
                cf_total = cf_total + cf * (1 - cf_total)
    return cf_total

# Hasil CF Flu
cf_flu = hitung_cf(gejala_user, pengetahuan["flu"])
print(f"CF diagnosis Flu setelah demam jadi 0.3 dan batuk 0.6 hasilnya = {cf_flu:.2f}")

CF diagnosis Flu setelah demam jadi 0.3 dan batuk 0.6 hasilnya = 0.72


# **2. Tambahkan 5 gejala baru beserta nilai CF**

In [33]:
# Gejala user (demam diubah menjadi 0.3 & batuk menjadi 0.6, dengan tambahan 5 gejala)
gejala_user = {
    "demam": 0.3,
    "batuk": 0.6,
    "sakit_tenggorokan": 0.6,
    "pilek": 0.7,
    "sakit_kepala": 0.6,
    "nyeri_otot": 0.4,
    "kelelahan": 0.5,
    "hidung_tersumbat": 0.3
}

# Pengetahuan pakar tetap (dengan tambahan 5 gejala)
pengetahuan = {
    "flu": {
        "demam": 0.8,
        "batuk": 0.7,
        "sakit_tenggorokan": 0.6,
        "pilek": 0.7,
        "sakit_kepala": 0.5,
        "nyeri_otot": 0.6,
        "kelelahan": 0.5,
        "hidung_tersumbat": 0.4
    }
}

# Fungsi perhitungan CF
def hitung_cf(gejala_user, pengetahuan_pakar):
    cf_total = 0
    first = True
    for gejala, cf_user in gejala_user.items():
        if gejala in pengetahuan_pakar:
            cf_pakar = pengetahuan_pakar[gejala]
            cf = cf_user * cf_pakar
            if first:
                cf_total = cf
                first = False
            else:
                cf_total = cf_total + cf * (1 - cf_total)
    return cf_total

# Hasil CF Flu
cf_flu = hitung_cf(gejala_user, pengetahuan["flu"])
print(f"CF diagnosis Flu setelah penambahan 5 gejala hasilnya = {cf_flu:.2f}")

CF diagnosis Flu setelah penambahan 5 gejala hasilnya = 0.95


# **3. Jika "demam" diubah jadi 0.2, bagaimana hasilnya? Mengapa?**

In [34]:
# Gejala user (demam diubah menjadi 0.2)
gejala_user = {
    "demam": 0.2,
    "batuk": 0.6,
    "sakit_tenggorokan": 0.6,
    "pilek": 0.7,
    "sakit_kepala": 0.6,
    "nyeri_otot": 0.4,
    "kelelahan": 0.5,
    "hidung_tersumbat": 0.3
}

# Pengetahuan pakar tetap (dengan tambahan 5 gejala)
pengetahuan = {
    "flu": {
        "demam": 0.8,
        "batuk": 0.7,
        "sakit_tenggorokan": 0.6,
        "pilek": 0.7,
        "sakit_kepala": 0.5,
        "nyeri_otot": 0.6,
        "kelelahan": 0.5,
        "hidung_tersumbat": 0.4
    }
}

# Fungsi perhitungan CF
def hitung_cf(gejala_user, pengetahuan_pakar):
    cf_total = 0
    first = True
    for gejala, cf_user in gejala_user.items():
        if gejala in pengetahuan_pakar:
            cf_pakar = pengetahuan_pakar[gejala]
            cf = cf_user * cf_pakar
            if first:
                cf_total = cf
                first = False
            else:
                cf_total = cf_total + cf * (1 - cf_total)
    return cf_total

# Hasil CF Flu
cf_flu = hitung_cf(gejala_user, pengetahuan["flu"])
print(f"CF diagnosis Flu setelah penambahan 5 gejala hasilnya = {cf_flu:.2f}")

CF diagnosis Flu setelah penambahan 5 gejala hasilnya = 0.94


**Gejala User (Dengan "demam" = 0.2)**

```python
gejala_user = {
    "demam": 0.2, #diubah dari asalnya 3
    "batuk": 0.6,
    "sakit_tenggorokan": 0.6,
    "pilek": 0.7,
    "sakit_kepala": 0.6,
    "nyeri_otot": 0.4,
    "kelelahan": 0.5,
    "hidung_tersumbat": 0.3
}

pengetahuan = {
    "flu": {
        "demam": 0.8,
        "batuk": 0.7,
        "sakit_tenggorokan": 0.6,
        "pilek": 0.7,
        "sakit_kepala": 0.5,
        "nyeri_otot": 0.6,
        "kelelahan": 0.5,
        "hidung_tersumbat": 0.4
    }
}
```

**Perhitungan Langkah per Langkah**

1. demam: 0.2 × 0.8 = 0.16
2. batuk: 0.6 × 0.7 = 0.42  
   `cf = 0.16 + 0.42 * (1 - 0.16) = 0.5128`
3. sakit_tenggorokan: 0.6 × 0.6 = 0.36  
   `cf = 0.5128 + 0.36 * (1 - 0.5128) = 0.6875`
4. pilek: 0.7 × 0.7 = 0.49  
   `cf = 0.6875 + 0.49 * (1 - 0.6875) = 0.8409`
5. sakit_kepala: 0.6 × 0.5 = 0.3  
   `cf = 0.8409 + 0.3 * (1 - 0.8409) = 0.8886`
6. nyeri_otot: 0.4 × 0.6 = 0.24  
   `cf = 0.8886 + 0.24 * (1 - 0.8886) = 0.9154`
7. kelelahan: 0.5 × 0.5 = 0.25  
   `cf = 0.9154 + 0.25 * (1 - 0.9154) = 0.9366`
8. hidung_tersumbat: 0.3 × 0.4 = 0.12  
   `cf = 0.9366 + 0.12 * (1 - 0.9366) = 0.9442`

**Hasil Akhir**

```python
CF diagnosis Flu ≈ 0.9442 -> Dibulatkan menjadi 0.94
```

**Kenapa Nilainya Menurun Saat "demam" Jadi 0.2?**

- Karena nilai awal CF lebih kecil, maka efek penguatan dari gejala selanjutnya juga lebih kecil.
- "Demam" punya bobot besar di pengetahuan pakar (0.8), jadi menurunkan nilainya mengurangi kekuatan akumulasi CF secara signifikan.
- Metode kombinasi CF sensitif terhadap urutan dan nilai awal gejala.

# **4. Jelaskan bagaimana hasil perubahannya  dalam bentuk komentar atau teks cell**
Penjelasan :
> Perubahan Data :
1. Nilai gejala diubah:
* demam dari 0.7 ➝ 0.3
* batuk dari 0.5 ➝ 0.6
* sakit_tenggorokan dari 0.4 ➝ 0.6
2. Penambahan 5 gejala baru:
* pilek: 0.7
* sakit_kepala: 0.6
* nyeri_otot: 0.4
* kelelahan: 0.5
* hidung_tersumbat: 0.3

>Dampak terhadap Perhitungan CF:
1. Semakin banyak gejala yang relevan dengan pengetahuan pakar, maka nilai CF total bisa meningkat — jika gejala tersebut memiliki kontribusi signifikan (nilai user dan pakar sama-sama tinggi).
2. Dalam diagnosa ini:
* Meskipun nilai demam (komponen besar sebelumnya) diturunkan, penambahan 5 gejala baru dengan nilai CF moderat memberikan kontribusi tambahan yang positif.
* Proses kombinasi CF menggunakan rumus:
       CF_total = CF1 + CF2 * (1 - CF1)

>Hasil Diagnosa:
* Sebelumnya (3 gejala): hasil CF kemungkinan lebih rendah, sekitar 0.72
* Setelah penambahan (8 gejala): hasil CF menjadi lebih tinggi karena lebih banyak gejala cocok, menjadi 0.95

